backend/app/consents.py

```diff
@@ -0,0 +1,184 @@
+    pass
+
+
+@lru_cache(maxsize=1)
+def get_key_fetcher() -> AsyncKeyFetcher:
+    """
+    Create and return a singleton instance of the AsyncKeyFetcher
+    """
+    return AsyncKeyFetcher(valid_issuers=[conf.OIDC_PROVIDER_URL])
+
+
+async def parse_token(id_token: Optional[str]) -> str:
+    """
+    Parse id_token, validate it and return a subject (user id)
+    :param id_token: ID Token from cookie
+    :return: Sub (user ID)
+    """
+    if not id_token:
+        raise HTTPException(401, "User not logged in")
+    try:
+        token = await validate_token(id_token=id_token)
+    except jwt.exceptions.InvalidTokenError:
+        raise HTTPException(401, "User not logged in")
+
+    return token["sub"]
+
+
+async def validate_token(id_token: str) -> dict:
+    """
+    Validate an id_token
+    raise: jwt.exceptions.InvalidTokenError If token is invalid
+    """
+    try:
+        key_entry = await get_key_fetcher().get_key(id_token)
+    except JWTKeyFetcherError as e:
+        raise jwt.exceptions.InvalidTokenError from e
+    return jwt.decode(
+        id_token,
+        verify=True,
+        audience=conf.OIDC_CLIENT_ID,
+        issuer=conf.OIDC_PROVIDER_URL,
+        **key_entry,
+    )
+
+
+def create_consent_request_token(sub) -> dict:
+    key = conf.PRIVATE_KEY
+
+    now = int(time.time())
+    crt = {
+        "iss": conf.CONSENT_REQUEST_TOKEN_ISSUER,
+        "sub": sub,
+        "subiss": conf.OIDC_PROVIDER_URL,
+        "acr": conf.OIDC_ACR_VALUES,
+        "app": conf.OIDC_CLIENT_ID,
+        "appiss": conf.OIDC_PROVIDER_URL,
+        "aud": conf.CONSENT_PORTAL_URL,
+        "exp": now + conf.CONSENT_REQUEST_TOKEN_VALID_SECONDS,
+        "iat": now,
+    }
+    headers = {
+        "v": "0.2",
+        "kid": key.kid,
+    }
+    token = jwt.encode(
+        payload=crt,
+        key=key.get_secret_value(),
+        algorithm=key.alg,
+        headers=headers,
+    )
+    return token
+
+
+def make_dsi_uri(definition: str) -> str:
+    """
+    Make Data Source Identifier
+    :param definition: Data Product Definition
+    :return: Data Source Identifier
+    """
+    return f"dpp://{conf.DATA_PRODUCT_SOURCE}@{conf.DATASPACE_DOMAIN}/{definition}"
+
+
+async def request_consent_token(dsi: str, sub: str) -> Optional[str]:
+    """
+    Fetch Consent Token from Consent Portal
+    :param dsi: Data Source Identifier
+    :param sub: Subject for a consent
+    :return: Consent Token if it has been granted already, None otherwise
+    """
+    url = f"{conf.CONSENT_PORTAL_URL}/Consent/GetToken"
+    cr_token = create_consent_request_token(sub)
+    headers = {"X-Consent-Request-Token": cr_token}
+    payload = {"dataSource": dsi}
+    async with httpx.AsyncClient() as client:
+        resp = await client.post(url, json=payload, headers=headers, timeout=30)
+    data = resp.json()
+    if data.get("type") == "consentGranted":
+        return data["consentToken"]
+    else:
+        raise MissingConsentToken
```

Comment:
Should we call it `MissingConsent` instead, as the reason for not having a consent token at hand is that the user has not granted the consent?